### PR TITLE
Add internal provider

### LINKED ISSUE #5374: Dashboard behind PathPrefix in 2.0 fails to load static assets
### Do you want to request a *feature* or report a *bug*?

Bug

### Did you try using a 1.7.x configuration for the version 2.0?

- [ ] Yes
- [x] No

### What did you do?

The new dashboard in 2.0 does not seem to handle being put behind a `PathPrefix`. More specifically, the frontend continues to look for assets in `/dashboard` instead of `/traefik/dashboard`. Configuration:

```toml
[http.routers]
    [http.routers.api]
        entryPoints = ["http"]
        rule = "HostRegexp(`systems.{tld:[a-z0-9.-]+}`) && PathPrefix(`/traefik`)"
        service = "api@internal"
        middlewares = ["adminAuth", "traefikAcp", "securityHeaders"]

[http.middlewares]
    [http.middlewares.traefikAcp.stripPrefix]
        prefixes = ["/traefik"]
```

Generated HTML:

```html
<script type="text/javascript" src="/dashboard/js/app.b1a5516b.js"></script>
<script type="text/javascript" src="/dashboard/js/vendor.bf460595.js"></script>
```

### What did you expect to see?

The new dashboard, i.e. assets being loaded from relative paths.

### What did you see instead?

A white page, because assets are loaded with an absolute path.

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
{"Version":"2.0.0-rc3","Codename":"montdor"}
```

### PR DESCRIPTION
**Pull request by ldez**
### What does this PR do?

- refactor core server.
- internal provider.

### Motivation

Better separation of concerns in the Traefik core.

Split `Server` into:
 - `ConfigurationWatcher`: Listens to provider and merges configuration in a configuration
 - `TCPEntryPoints`: Allows operations on all TCPEntryPoint (start, switch, stop)
 - `TCPRouterFactory`: Creates the TCP routers per entry point from configuration

Remove `RouteAppenderFactory`, replaced by the internal provider. 

Fixes #5374

### More

- [x] Added/updated tests
- [x] Added/updated documentation

### Additional Notes

Co-authored-by: Julien Salleyron <julien.salleyron@gmail.com>

### CODE REVIEW

File: pkg/server/middleware/chainbuilder.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+	"github.com/containous/traefik/v2/pkg/metrics"
+	"github.com/containous/traefik/v2/pkg/middlewares/accesslog"
+	metricsmiddleware "github.com/containous/traefik/v2/pkg/middlewares/metrics"
+	"github.com/containous/traefik/v2/pkg/middlewares/requestdecorator"
+	mTracing "github.com/containous/traefik/v2/pkg/middlewares/tracing"
+	"github.com/containous/traefik/v2/pkg/tracing"
+	"github.com/containous/traefik/v2/pkg/tracing/jaeger"
+)
+
+// ChainBuilder Creates a middleware chain by entry point.
```
Comment (mpl) [inline pkg/server/middleware/chainbuilder.go]:
```suggestion
// ChainBuilder Creates a middleware chain by entry point. It is used for middlewares that are created almost systematically and that need to be created before all others.
```

File: pkg/server/middleware/chainbuilder.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+type ChainBuilder struct {
+	metricsRegistry        metrics.Registry
+	accessLoggerMiddleware *accesslog.Handler
+	tracer                 *tracing.Tracing
+	requestDecorator       *requestdecorator.RequestDecorator
+}
+
+// NewChainBuilder Creates a mew ChainBuilder.
```
Comment (mpl) [inline pkg/server/middleware/chainbuilder.go]:
```suggestion
// NewChainBuilder Creates a new ChainBuilder.
```

File: pkg/server/middleware/chainbuilder.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+	if conf.Haystack != nil {
+		if backend != nil {
+			log.WithoutContext().Error("Multiple tracing backend are not supported: cannot create Haystack backend.")
+		} else {
+			backend = conf.Haystack
+		}
+	}
+
+	if backend == nil {
+		log.WithoutContext().Debug("Could not initialize tracing, use Jaeger by default")
```
Comment (mpl) [inline pkg/server/middleware/chainbuilder.go]:
```suggestion
		log.WithoutContext().Debug("Could not initialize tracing, using Jaeger by default")
```

File: integration/access_log_test.go [hunk 4]
```diff
@@ -580,28 +589,31 @@ func CheckAccessLogFormat(c *check.C, line string, i int) {
...
-	c.Assert(results[accesslog.RequestCount], checker.Equals, fmt.Sprintf("%d", i+1))
+	count, _ := strconv.Atoi(results[accesslog.RequestCount])
+	c.Assert(count, checker.GreaterOrEqualThan, i+1)
 	c.Assert(results[accesslog.RouterName], checker.Matches, `^"?`+v.routerName+`.*(@docker)?$`)
 	c.Assert(results[accesslog.ServiceURL], checker.Matches, `^"?`+v.serviceURL+`.*$`)
 	c.Assert(results[accesslog.Duration], checker.Matches, `^\d+ms$`)
 }
 
 func waitForTraefik(c *check.C, containerName string) {
+	time.Sleep(1 * time.Second)
```
[the remark this answers is not in the document]
Comment (ldez) [reply integration/access_log_test.go]: The retry L621 is not a wait.
We need to wait 1s before trying.

File: pkg/server/configurationwatcher.go [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+		logger.Debugf("Unable to marshal provider configuration %T: %v", c.provider, err)
+	}
+
+	logger.Infof("Starting provider %T %s", c.provider, jsonConf)
+	currentProvider := c.provider
+
+	safe.Go(func() {
+		err := currentProvider.Provide(c.configurationChan, c.routinesPool)
+		if err != nil {
+			logger.Errorf("Error starting provider %T: %s", c.provider, err)
```
[the remark this answers is not in the document]
Comment (ldez) [reply pkg/server/configurationwatcher.go]: there is a closure effect here, we need to use the variable.

**Pull request merged**